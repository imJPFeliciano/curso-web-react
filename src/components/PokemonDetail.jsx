import React, { useState, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { PokemonCardContainer, PokemonName, PokemonImage } from './Pokemon.styled';
import styled from 'styled-components';
import { getPokemonImage } from '../utils/pokemonUtils';
import TypeChip from './PokemonTypeChip';

const BASE_URL = 'https://pokeapi.co/api/v2/pokemon';

const PokemonDetailContainer = styled.div`
`;

const renderTypes = ({ types }) => {
  return types.map(pokemonType => {
    return (
      <TypeChip
        key={pokemonType.type.name}
        name={pokemonType.type.name}
      />
    );
  });
}

const PokemonDetail = ({ loading, pokemon, image }) => {
  return (
    <PokemonDetailContainer>
      { loading && <span className="p-loader">loading ...</span>}
      { !loading && pokemon && (
          <PokemonCardContainer>
            <PokemonImage src={image} />
            <PokemonName>{pokemon.id}: {pokemon.name}</PokemonName>
            { pokemon && pokemon.types && renderTypes({ types: pokemon.types }) }
          </PokemonCardContainer>
        )
      }
    </PokemonDetailContainer>
  );
};

const PokemonDetailWrapper = () => {
  const { id } = useParams();
  // TODO: validar que sea un numero correcto (intervalo de pokemons)

  const [pokemon, setPokemon] = useState(null);
  const [pokemonId] = useState(Number(id));
  const [loading, setLoading] = useState(true);

  const pokemonImage = getPokemonImage(id);

  useEffect(() => {
    fetch(`${BASE_URL}/${pokemonId}`)
      .then(response => {
        return response.json()
      })
      .then(data => {
        setPokemon(data);
        setLoading(false);
      })
  }, [])

  return (
    <PokemonDetail loading={loading} pokemon={pokemon} image={pokemonImage} />
  )
}

export default PokemonDetailWrapper;
export {
  PokemonDetail,
};
