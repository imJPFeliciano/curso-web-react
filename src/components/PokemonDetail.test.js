import React from 'react';
import { PokemonDetail } from './PokemonDetail';
import { PokemonName, PokemonImage } from './Pokemon.styled';

import { mount, shallow } from 'enzyme';

const props = {
  loading: true,
  pokemon: {
    id: 1,
    name: 'Bulbasaur',
    types: [
      { type: { name: 'Grass' } },
    ]
  },
  imagen: 'someURL',
};

describe('Pokemon Detail', () => {
  it('render sin problemas', () => {
    const wrapper = mount(<PokemonDetail {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('muestra el mensaje de cargando', () => {
    const wrapper = shallow(<PokemonDetail {...props} />);
    expect(wrapper.find('.p-loader')).toHaveLength(1);
  });

  it('muestra la informacion del pokemon', () => {
    const mockProps = {
      ...props,
      loading: false,
    };

    const wrapper = shallow(<PokemonDetail {...mockProps} />);
    expect(wrapper.find(PokemonImage)).toHaveLength(1);
    expect(wrapper.find(PokemonName)).toHaveLength(1);
  });
});