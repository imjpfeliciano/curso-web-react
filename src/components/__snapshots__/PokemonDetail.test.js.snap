// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Detail render sin problemas 1`] = `ReactWrapper {}`;
